### PR TITLE
Add `Typing/MethodReturnTypeRestriction`

### PR DESCRIPTION
**Pull request by nobodywasishere**

### CODE REVIEW

File: spec/ameba/rule/typing/method_return_type_restriction_spec.cr [hunk 1]
```diff
@@ -0,0 +1,117 @@
+require "../../../spec_helper"
+
+module Ameba::Rule::Typing
+  subject = MethodReturnTypeRestriction.new
+  subject.private_methods = true
+  subject.protected_methods = true
+  subject.undocumented = true
```
Comment (veelenga) [inline spec/ameba/rule/typing/method_return_type_restriction_spec.cr:7]: at top level we usually test the default config, then the props can be toggled in a separate context.

**Review by veelenga — APPROVED** [review]
lgtm

File: spec/ameba/rule/typing/method_return_type_restriction_spec.cr [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+    expect_no_issues subject, <<-CRYSTAL
+      def hello : String
+        "hello world"
+      end
+
+      # This method is documented
+      def hello : String
+        "hello world"
+      end
+
```
Comment (Sija) [inline spec/ameba/rule/typing/method_return_type_restriction_spec.cr:16]: That example seems redundant.

```suggestion
```

File: spec/ameba/rule/typing/method_return_type_restriction_spec.cr [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+      protected def hello : String
+        "hello world"
+      end
+
+      # :nodoc:
+      def hello : String
+        "hello world"
+      end
```
Comment (Sija) [inline spec/ameba/rule/typing/method_return_type_restriction_spec.cr:28]: ditto

```suggestion
```

File: spec/ameba/rule/typing/method_return_type_restriction_spec.cr [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+
+          # This method is documented
+          def hello : String
+            "hello world"
+          end
+
+          private def hello : String
+            "hello world"
+          end
+
```
Comment (Sija) [inline spec/ameba/rule/typing/method_return_type_restriction_spec.cr:146]: ditto

```suggestion
```

File: spec/ameba/rule/typing/method_return_type_restriction_spec.cr [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+          protected def hello : String
+            "hello world"
+          end
+
+          # :nodoc:
+          def hello : String
+            "hello world"
+          end
```
Comment (Sija) [inline spec/ameba/rule/typing/method_return_type_restriction_spec.cr:154]:
```suggestion
```

File: src/ameba/rule/typing/method_return_type_restriction.cr [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+  class MethodReturnTypeRestriction < Base
+    properties do
+      description "Recommends that methods have a return type restriction"
+      enabled false
+      undocumented false
+      private_methods false
+      protected_methods false
+    end
+
+    MSG = "Methods should have a return type restriction"
```
Comment (Sija) [inline src/ameba/rule/typing/method_return_type_restriction.cr:51]:
```suggestion
    MSG = "Method should have a return type restriction"
```

File: src/ameba/rule/typing/method_return_type_restriction.cr [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+    def test(source, node : Crystal::Def)
+      return if node.return_type || check_config(node)
+
+      issue_for node, MSG, prefer_name_location: true
+    end
+
+    def check_config(node : Crystal::ASTNode) : Bool
```
Comment (Sija) [inline src/ameba/rule/typing/method_return_type_restriction.cr:59]:
```suggestion
    def valid?(node : Crystal::ASTNode) : Bool
```

File: src/ameba/rule/typing/method_return_type_restriction.cr [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+  #   Enabled: false
+  #   Undocumented: false
+  #   PrivateMethods: false
+  #   ProtectedMethods: false
+  # ```
+  class MethodReturnTypeRestriction < Base
+    properties do
+      description "Recommends that methods have a return type restriction"
+      enabled false
+      undocumented false
```
Comment (Sija) [inline src/ameba/rule/typing/method_return_type_restriction.cr:46]: This option doesn't seem to belong here (as in other PRs).

**Review by Sija — CHANGES_REQUESTED** [review]

File: spec/ameba/rule/typing/method_return_type_restriction_spec.cr [hunk 9]
```diff
@@ -0,0 +1,127 @@
+require "../../../spec_helper"
+
+module Ameba::Rule::Typing
+  subject = MethodReturnTypeRestriction.new
+
+  it "passes if a method has a return type" do
```
Comment (Sija) [inline spec/ameba/rule/typing/method_return_type_restriction_spec.cr:6]:
```suggestion
  it "passes if a method has a return type restriction" do
```

ditto elsewhere

File: spec/ameba/rule/typing/method_return_type_restriction_spec.cr [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+      CRYSTAL
+  end
+
+  it "passes if a private or protected method doesn't have a return type" do
+    expect_no_issues subject, <<-CRYSTAL
+      private def hello
+        "hello world"
+      end
+
+      protected def hello : String
```
Comment (Sija) [inline spec/ameba/rule/typing/method_return_type_restriction_spec.cr:28]:
```suggestion
      protected def hello
```

File: src/ameba/rule/typing/method_return_type_restriction.cr [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+  #
+  # ```
+  # def hello
+  #   "hello world"
+  # end
+  #
+  # def listen(a, b)
+  #   0
+  # end
+  # ```
```
Comment (Sija) [inline src/ameba/rule/typing/method_return_type_restriction.cr:26]: Please, keep the examples minimal, functional and list the invalid cases first.

```suggestion
  # For example, this are considered invalid:
  #
  # ```
  # def hello(name = "World")
  #   "Hello #{name}"
  # end
  # ```
  #
  # And this is valid:
  #
  # ```
  # def hello(name = "World") : String
  #   "Hello #{name}"
  # end
  # ```
```

File: src/ameba/rule/typing/method_return_type_restriction.cr [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+  # YAML configuration example:
+  #
+  # ```
+  # Typing/MethodReturnTypeRestriction:
+  #   Enabled: false
+  #   PrivateMethods: false
+  #   ProtectedMethods: false
+  # ```
+  class MethodReturnTypeRestriction < Base
+    properties do
```
Comment (Sija) [inline src/ameba/rule/typing/method_return_type_restriction.cr:40]:
```suggestion
    properties do
      since_version "1.7.0"
```

File: src/ameba/rule/typing/method_return_type_restriction.cr [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+      private_methods false
+      protected_methods false
+    end
+
+    MSG = "Method should have a return type restriction"
+
+    def test(source, node : Crystal::Def)
+      return if node.return_type || valid?(node)
+
+      issue_for node, MSG, prefer_name_location: true
```
Comment (Sija) [inline src/ameba/rule/typing/method_return_type_restriction.cr:52]: This rule relates to the whole method definition, not just the name.

```suggestion
      issue_for node, MSG
```

**Review by Sija — CHANGES_REQUESTED** [review]

File: spec/ameba/rule/typing/method_return_type_restriction_spec.cr [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+        expect_no_issues rule, <<-CRYSTAL
+          def hello : String
+            "hello world"
+          end
+
+          # This method is documented
+          def hello : String
+            "hello world"
+          end
+
```
Comment (Sija) [inline spec/ameba/rule/typing/method_return_type_restriction_spec.cr:58]:
```suggestion
```

File: src/ameba/rule/typing/method_return_type_restriction.cr [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+      return if node.return_type || valid?(node)
+
+      issue_for node, MSG
+    end
+
+    def valid?(node : Crystal::ASTNode) : Bool
+      (!private_methods? && node.visibility.private?) ||
+        (!protected_methods? && node.visibility.protected?) ||
+        false
+    end
```
Comment (Sija) [inline src/ameba/rule/typing/method_return_type_restriction.cr:52]:
```suggestion
    def valid_return_type?(node : Crystal::ASTNode) : Bool
      !!node.return_type ||
        (node.visibility.private? && !private_methods?) ||
          (node.visibility.protected? && !protected_methods?)
    end
```

File: src/ameba/rule/typing/method_return_type_restriction.cr [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+      private_methods false
+      protected_methods false
+    end
+
+    MSG = "Method should have a return type restriction"
+
+    def test(source, node : Crystal::Def)
+      return if node.return_type || valid?(node)
+
+      issue_for node, MSG
```
Comment (Sija) [inline src/ameba/rule/typing/method_return_type_restriction.cr:45]:
```suggestion
      issue_for node, MSG unless valid_return_type?(node)
```

**Review by Sija — CHANGES_REQUESTED** [review]

File: src/ameba/rule/typing/method_return_type_restriction.cr [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+    MSG = "Method should have a return type restriction"
+
+    def test(source, node : Crystal::Def)
+      issue_for node, MSG unless valid_return_type?(node)
+    end
+
+    def valid_return_type?(node : Crystal::ASTNode) : Bool
+      !!node.return_type ||
+        (node.visibility.private? && !private_methods?) ||
+          (node.visibility.protected? && !protected_methods?)
```
Comment (Sija) [inline src/ameba/rule/typing/method_return_type_restriction.cr:49]:
```suggestion
        (node.visibility.protected? && !protected_methods?)
```

**Review by Sija — APPROVED** [review]

**Pull request merged**